astropy/modeling/powerlaws.py

```diff
@@ -583,27 +583,26 @@ class Schechter1D(Fittable1DModel):
 
     phi_star = Parameter(default=1., description=('Normalization factor '
                                                   'in units of number density'))
-    m_star = Parameter(default=-20., description='Characteristic magnitude')
+    m_star = Parameter(default=-20., description='Characteristic magnitude', mag=True)
     alpha = Parameter(default=-1., description='Faint-end slope')
 
     @staticmethod
     def evaluate(mag, phi_star, m_star, alpha):
         """Schechter luminosity function model function."""
-        if isinstance(mag, Quantity) or isinstance(m_star, Quantity):
-            raise ValueError('mag and m_star must not have units')
-        factor = 10 ** (0.4 * (m_star - mag))
 
-        return (0.4 * np.log(10) * phi_star * factor**(alpha + 1)
-                * np.exp(-factor))
+        factor_exp = 0.4 * (m_star - mag)
```

Comment:
There is no check that `m_star` and `mag` are in the same magnitude *system*?

Reply:
Shouldn't that fail or raise a unit error?

Reply:
🤷

Reply:
```python
import astropy.units as u

a = 5 * u.ABmag
b = 3 * u.STmag

print(a - b)
```
returns
```python
2.0 mag(AB / ST
```

Reply:
@larrybradley do `m_star ` and `mag` need to be in the same magnitude system?

Reply:
That hurts my head... 😆 I'll defer to @larrybradley and @mhvk .

Reply:
One should not be able to mix different magnitude systems because they have different zero points (and in some cases meanings, e.g. Vega mag vs. AB mag).  I don't know what a unit of `mag(AB / ST)` even means.  I would have expected an error.

Reply:
Alright I'll add a check to ensure that `m_star` and `mag` have the same units

Reply:
Note that `m_star` and `mag` must be in magnitude units.  In the current implementation, someone could enter them both in arbitrary units (e.g., Jy), which would not raise an error, but also does not make sense.

In my original PR supporting mag units, I wrote a helper method to deal with the input units:
* `m_star` and `mag` must have identical units and they must be magnitude units
* 
https://github.com/larrybradley/astropy/blob/schechter-mag/astropy/modeling/powerlaws.py#L590

Reply:
In principle, the magnitudes should evaluate correctly. E.g.,
```
m_star = -20 * u.ABmag
mag = -22.5 * u.ABmag
factor = (m_star - mag).to(u.dimensionless_unscaled)
factor
# <Quantity 0.1>
```
And of course this will fail if the magnitudes are not in the same system.

Might it be easiest to just convert the inputs to `Magnitude` for this case?

Reply:
p.s. Incompatible magnitude units like `AB` and `ST` would raise an error if one not just subtracts, but tries to convert to dimensionless.

Reply:
Hey this is taken care of by the `input_units` method.